Adjust GitHub and Jira extraction to a consistent format

fehnkera opened the pull request:
'issue_processing.py' and 'jira_issue_processing.py' get adjusted, so
they both can print a '.list' file with a consistent format. The new
format is:
"issue number";"issue name";"[issue types]";"issue state";"issue creation
date";"issue resolve date";"[issue components]";"event type";"event user
name";"event user e-mail";"event time";"event info 1";"event info 2"

Attached diagram gives a overview about how the extraction to the new format works:
[codeface-data-extraction-diagramm.pdf](https://github.com/se-passau/codeface-extraction/files/2265443/codeface-data-extraction-diagramm.pdf)

For now only the methods 'print_to_disk_new' ('issue_processing.py') and
'print_to_disk_bugs' ('jira_issue_processing.py') use this new format,
because the network library only supports the old format. When the
network library is updated all output methods shall use this format.

This fixes #91.

Signed-off-by: Anselm Fehnker <fehnker@fim.uni-passau.de>

Commit:
Adjust jira and github issue processing to a consistent format

Adjust 'issue_processing.py' and 'jira_issue_processing.py', so
they both can print a '.list' file with a consistent format. The new
format is:
- issue number
- list of issue types
- issue state
- issue creation date
- issue resolve date
- list of issue components
- event type
- event user name
- event user e-mail
- event time
- event info 1
- event info 2
Only the methods 'print_to_disk_new' ('issue_processing.py') and
'print_to_disk_bugs' (jira_issue_processing.py') use this new format,
because the network library still works with the old format. When the
network library is updated all output methods shall use this format.

Signed-off-by: Anselm Fehnker <fehnker@fim.uni-passau.de>

Commit:
Add "issue title" to new '.list' format

Extends the  new '.list' file format from previous commit with the
issue title. Now the new format is:
- issue number
- issue title
- list of issue types
- issue state
- issue creation date
- issue resolve date
- list of issue components
- event type
- event user name
- event user e-mail
- event time
- event info 1
- event info 2

Signed-off-by: Anselm Fehnker <fehnker@fim.uni-passau.de>

Commit:
Update comments and reformat code before pull-request

Add and updates few comments. Remove redundant blank lines and
formattes the code

Signed-off-by: Anselm Fehnker <fehnker@fim.uni-passau.de>

ecklbarb reviewed (changes requested):
Thank you for this nice additions,  @fehnkera!

ecklbarb commented on issue_processing/issue_processing.py:
```diff
@@ -96,28 +100,149 @@ def reformat(issue_data):
+
+        # if an issue has no eventsList an empty Lists gets created
+        if issue['eventsList'] is None:
+            issue['eventsList'] = []
```
Please use consistent quotation marks.

ecklbarb commented on issue_processing/issue_processing.py:
```diff
@@ -96,28 +100,149 @@ def reformat(issue_data):
+        issue['created_at'] = format_time(issue['created_at'])
+
+        # parses the close time in the correct format
+        issue['closed_at'] = format_time(issue['closed_at'])
```
See the comment above.

ecklbarb commented on issue_processing/issue_processing.py:
```diff
@@ -96,28 +100,149 @@ def reformat(issue_data):
+                    return d.strftime('%Y-%m-%d %H:%M:%S')
+                except ValueError:
+                    # if the given time format is not know it can't be formatted, so it just gets returned
+                    # also happens if the time already has the correct format (old GitHub-Wrapper files)
```
Can you check as first step, whether the date has the correct format?

ecklbarb commented on issue_processing/issue_processing.py:
```diff
@@ -96,28 +100,149 @@ def reformat(issue_data):
+            # format from new GitHub-Wrapper files
+            d = datetime.strptime(time, '%Y-%m-%dT%H:%M:%SZ')
+            return d.strftime('%Y-%m-%d %H:%M:%S')
+        except ValueError:
```
Maybe, you can find an alternative to the try-except.

clhunsen replied:
Instead of the `try-except` cascade, we may try something like [this](https://stackoverflow.com/a/14245142):
```python
from dateutil import parser as dateparser
dateparser.parse(string)
```

Otherwise, there is a suggestion in [another answer of the same question](https://stackoverflow.com/a/14245134) on how to use a loop instead of the current `try-except` cascade (maybe with some more documentation :wink:):
```python
for date_format in DATE_FORMATS:
    try:
        my_date = datetime.strptime(test_date, date_format)
    except ValueError:
        pass
    else:
      break
else:
  my_date = None
```

fehnkera replied:
The dateutil parser works. Thank you very much for the suggestion @clhunsen

ecklbarb commented on issue_processing/issue_processing.py:
```diff
@@ -96,28 +100,149 @@ def reformat(issue_data):
+
+        # updates format of every event in the eventsList of an issue
         for event in issue["eventsList"]:
             event["ref_target"] = ""
```
I think, the code would be clearer, if you extract the parsing of the event information in a separate method.

clhunsen replied:
Maybe, we should do that also for the information on related issues and related commits.

ecklbarb commented on issue_processing/issue_processing.py:
```diff
@@ -128,17 +253,107 @@ def reformat(issue_data):
+
+        # state and resolution change lists get sorted by time
+        state_changes.sort(key=lambda x: x[0])
+        resolution_changes.sort(key=lambda x: x[0])
```
Please use empty spaces before and after `=`.

ecklbarb commented on issue_processing/issue_processing.py:
```diff
@@ -128,17 +253,107 @@ def reformat(issue_data):
                                not (event["user"] is None or event["ref_target"] is None)]
 
+        # sorts eventsList by time
+        issue["eventsList"].sort(key=operator.itemgetter('created_at'))
```
Please, use empty spaces before and after `=` and consistent quotation marks.

clhunsen replied:
Before I look at the details of this PR myself, one small comment and advice here: Python has a style guide ([PEP8](https://www.python.org/dev/peps/pep-0008/); unlike R), which is also directly supported by PyCharm. You just need to re-format the file (press <kbd>Alt+Shift+L</kbd>).
I highly recommend using PyCharm's formatter. This will solve most of the formatting issues up-front (except for quotation marks, I suppose).

Why I comment here: The style guide states the following – in contrast to our coding conventions from the network library:
> Don't use spaces around the = sign when used to indicate a keyword argument or a default parameter value.

ecklbarb replied:
Thank you for this hint @clhunsen.

ecklbarb commented on issue_processing/issue_processing.py:
```diff
@@ -208,12 +423,16 @@ def get_or_update_user(user, buffer_db=user_buffer):
 
 def print_to_disk(issues, results_folder):
-    """Print issues to file 'issues.list' in result folder
+    """Print issues to file 'issues.list' in result folder.
```
Please, make a new line after `"""`.

ecklbarb commented on issue_processing/issue_processing.py:
```diff
@@ -208,12 +423,16 @@ def get_or_update_user(user, buffer_db=user_buffer):
+    """Print issues to file 'issues.list' in result folder.
+    This format is outdated but still used by the network library.
+    When the network library is updated, this method can be overwritten by 'print_to_disk_new'.
 
```
Good hint. Thank you for this comment.

clhunsen replied:
Please add a `TODO` in front of this line, so we can easily re-find this comment.

ecklbarb commented on issue_processing/issue_processing.py:
```diff
@@ -242,3 +461,41 @@ def print_to_disk(issues, results_folder):
+
+
+def print_to_disk_new(issues, results_folder):
+    """Print issues to file 'issues_new.list' in result folder.
```
` """`
`Print issues to file 'issues_new.list' in result folder.`

ecklbarb commented on issue_processing/jira_issue_processing.py:
```diff
@@ -178,30 +178,56 @@ def parse_xml(issue_data, persons):
+                    history['event_info_1'] = ref.getElementsByTagName('issuekey')[0].firstChild.data
+                    history['event_info_2'] = "issue"
+
+                    issue['history'].append(history)
```
Please, use consistent quotation marks.

ecklbarb replied:
@clhunsen, what do you think: Which quotations marks should be use to be consistent?

clhunsen replied:
Does not matter to me as long as they are consistent in a single file – it makes reading the source code easier. Personally, I prefer double quotation marks.

fehnkera replied:
The problem with the not consistent quotation marks is in the whole file, not only the parts of the file that I worked on. Shall I also change the quotation marks in the other methods?

clhunsen replied:
Yes, please. A quick search-replace call will probably suffice.

ecklbarb reviewed (commented):
I commented on many small stuff. Basically, your changes look really fine.

clhunsen commented on issue_processing/issue_processing.py:
```diff
@@ -96,28 +100,149 @@ def reformat(issue_data):
+    # known resolutions from JIRA and GitHub default labels
+    known_resolutions = {"unresolved", "fixed", "wontfix", "duplicate", "invalid", "incomplete",
+                         "cannot reproduce", "later", "not a problem", "implemented", "done",
+                         "auto closed",
```
The line breaks here look weird. If you have some grouping here, please insert appropriate comments; otherwise, just re-flow the line breaks.

clhunsen commented on issue_processing/issue_processing.py:
```diff
@@ -128,17 +253,107 @@ def reformat(issue_data):
+            # parses the event time in the correct format
+            event["created_at"] = format_time(event["created_at"])
+
+            # formats the event according to it's event type
```
its

clhunsen commented on issue_processing/issue_processing.py:
```diff
@@ -96,28 +100,149 @@ def reformat(issue_data):
+    known_types = {"bug", "improvement", "enhancement", "new feature", "task", "test", "wish"}
+
+    # known resolutions from JIRA and GitHub default labels
+    known_resolutions = {"unresolved", "fixed", "wontfix", "duplicate", "invalid", "incomplete",
```
`known_resolutions` and also `known_types` are predestined to be global variables. Please move them outside of this function.

clhunsen commented on issue_processing/issue_processing.py:
```diff
@@ -128,17 +253,107 @@ def reformat(issue_data):
+                    # creates new resolution change object
+                    resolution_changes.append([resolution_event["created_at"], new_resolutions])
+
+            # %TODO: event for removed labels has to be checked. Data is missing
```
In the last data file I sent you, there is an event called `unlabeled`.

clhunsen commented on issue_processing/issue_processing.py:
```diff
@@ -242,3 +461,41 @@ def print_to_disk(issues, results_folder):
+
+def print_to_disk_new(issues, results_folder):
+    """Print issues to file 'issues_new.list' in result folder.
+    This file has a consistent format to the 'bugs-jira.list' file.
```
Please reference the function here that writes this particular file.

clhunsen commented on issue_processing/issue_processing.py:
```diff
@@ -242,3 +461,41 @@ def print_to_disk(issues, results_folder):
+def print_to_disk_new(issues, results_folder):
+    """Print issues to file 'issues_new.list' in result folder.
+    This file has a consistent format to the 'bugs-jira.list' file.
+    When the network library is updated, this is the format which shall be used.
```
Please add a `TODO` in front of this line, so we can easily re-find this comment.

clhunsen commented on issue_processing/issue_processing.py:
```diff
@@ -242,3 +461,41 @@ def print_to_disk(issues, results_folder):
+    When the network library is updated, this is the format which shall be used.
+
+    :param issues: the issues to dump
+    :param results_folder: the folder where to place 'issues.list' output file
```
`issues_new.list` :wink:

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -275,54 +301,140 @@ def load_issue_via_api(issues, persons, url):
     for issue in issues:
 
-        last_checked = issue['creationDate']  # needed to update state and resolution on creation for comments
+        def format_time(time):
```
Please move this function to the global level as it is a utility function.

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -432,12 +544,13 @@ def print_to_disk(issues, results_folder):
-def print_to_disk_bugs(issues, results_folder, skip_history):
+def print_to_disk_bugs(issues, results_folder):
     """Sorts of bug issues and prints them to file 'bugs-jira.list' in result folder
+    This method prints in a new format which is consistent to the new GitHub format.
```
What is "the new GitHub format"? :wink:  Please refer to the corresponding function in the other file.

clhunsen reviewed (changes requested):
Thank you very much for your work, @fehnkera! Overall, this looks terrific. :grinning: 

Unfortunately, there are some things to improve before we can merge (rather structural stuff, no  functional issues):
- Consider [this comment](https://github.com/se-passau/codeface-extraction/pull/16/files#r210663617) and reformat all changed files.
- Work on the other comments by Barbara and me.
- Rebase your commits and rewrite the commit messages for better structure (especially the commit titles, see [here](https://github.com/se-passau/codeface-extraction-r/blob/8c2a8255966cc6c38ea16e64a28b135ef8456e58/CONTRIBUTING.md#commit-messages) for recommendations). Using a more consistent formatting really helps with debugging in the future and also helps yourself to structure your changes and thoughts.

    For example, you can easily use an itemization in commit 9602adbec8012abfb54b6a2d0dedcce855e29805. This way, the columns in the new format can be parsed better by the reader.
- Please add the PDF file (and also the GRAPHML source file) for the diagram to the repository.

If you have any questions or need help, please do not hesitate to ask!

clhunsen commented:
> Regarding the last point to change, @bockthom and I asked @fehnkera not to add the diagram to the repository, but to include it in the pull request.

Ah, I have not known about that. Okay, then forget about the last point in my review, @fehnkera. Maybe, we find another (local) place (in the NFS) where to put it for later use.

Commit:
Apply review and split 'reformat' method

Applies review from @clhunsen and @ecklbarb. Reformat the code and uses
consistent quotation marks. The method 'reformat' is splitet in three
methods:
- 'reformat_issues': ensures that all issues have the same format
- 'merge_issue_events': puts all issue events in a similar format and
merges them into one list
- 'reformat_events': formats event information according to the event
type

Signed-off-by: Anselm Fehnker <fehnker@fim.uni-passau.de>

bockthom commented on issue_processing/issue_processing.py:
```diff
@@ -98,26 +129,133 @@ def reformat(issue_data):
+        # empty container for issue resolutions
+        issue["resolution"] = []
+
+        # if an issue has no eventsList an empty Lists gets created
```
Is there a reason for the plural in `empty Lists`? If not, could you please use the singular here?
In addition, adding a comma after the if-clause (i.e., after `eventsList`) will optimize the flow of reading.

This comment applies also for all the following comments regarding empty lists below.

bockthom commented on issue_processing/issue_processing.py:
```diff
@@ -98,26 +129,133 @@ def reformat(issue_data):
+        issue["state_new"] = "open"
+
+        # adds events for related issues to eventsList
+        # GH-Wrapper doesn't provide information about time and user yet. Has to be updated if this changes
```
Meanwhile, the GH-Wrapper does provide the information about time and user of related issues (it is not merged to master yet, but we already provided some example data on the new format, see here: https://github.com/se-passau/GitHubWrapper/issues/22#issuecomment-411324641
So, please include this (have a look at the most recent example data we have sent you).

bockthom commented on issue_processing/issue_processing.py:
```diff
@@ -98,26 +129,133 @@ def reformat(issue_data):
+            link_event["ref_target"] = ""
+            issue["eventsList"].append(link_event)
+
+        # adds events for related commits to eventsList
```
Please differentiate between referenced commits and commits which are part of a pull request:
- "relatedCommits": have `Username` (and name, email) and `Referenced_at` in the GH
- "hasCommit": `Username` (and name, email) and `Referenced_at` are `null` in the GH data, leading to empty strings for `user` and `created_at` here.

bockthom commented on issue_processing/issue_processing.py:
```diff
@@ -131,14 +269,83 @@ def reformat(issue_data):
+                    resolution_event["ref_target"] = ""
+                    issue["eventsList"].append(resolution_event)
+
+            # %TODO: event for removed labels has to be checked. Data is missing
```
There should already be examples for `unlabeled` in the most recent GH data. If not, let us now.

bockthom commented on issue_processing/jira_issue_processing.py:
```diff
@@ -178,30 +178,56 @@ def parse_xml(issue_data, persons):
 
             type = issue_x.getElementsByTagName('type')[0]
             issue['type'] = type.firstChild.data
+            # new consistent format with GitHub issues. Not supported by the network library yet
```
maybe also add a `TODO` here as the `new` parts should replace the current ones later, I guess. (Or I am wrong here?)

This applies also for the following comments below.

bockthom reviewed (changes requested):
Thanks for all the nice changes, @fehnkera. I really like the new format very well!

I just have a few comments regarding your changes, please have a look at them. (Some comments address the new format regarding referencing issues, referenced commits, and "unlabeled" events. I am not sure if you are currently working on them or just have forgotten about them -- anyway, just take my remarks on them as a friendly reminder.)

In addition, something went wrong with the rebase of the commits: Now, there are some commits not authored by you appearing in this pull-request. Those commits are already part of the master branch and should not be added again. So, please rebase again on the current version of master and remove the duplicated commits introduced by you. If you need help with the rebase, no worries, just let us know.

bockthom commented:
> > Regarding the last point to change, @bockthom and I asked @fehnkera not to add the diagram to the repository, but to include it in the pull request.
> 
> Ah, I have not known about that. Okay, then forget about the last point in my review, @fehnkera. Maybe, we find another (local) place (in the NFS) where to put it for later use.

Yes, I would prefer a local place. We can decide upon that when everything related to this PR is finished.

@fehnkera Could you please also update the diagram? If I remember correctly, the most recent changes are not yet updated in the diagram.

clhunsen commented:
> In addition, something went wrong with the rebase of the commits: Now, there are some commits not authored by you appearing in this pull-request. Those commits are already part of the master branch and should not be added again. So, please rebase again on the current version of master and remove the duplicated commits introduced by you. If you need help with the rebase, no worries, just let us know.

On this, @fehnkera sent an e-mail already, I will take care of this.

Commit:
Add method 'create_user' and update 'format_time'

Adds a method that creates a user object with attributes 'name',
'username' and 'email'. Now several methods can use 'create_user'
instead of manually creating the user object.
Furthermore the format_time method is updated, so it can also handle an
empty time correctly.

Signed-off-by: Anselm Fehnker <fehnker@fim.uni-passau.de>

Commit:
Add None case to 'format_time' method

The format time method in 'issue_processing' and 'jira_issue_processing'
now also proofs the none case as it could leed to errors before.

Signed-off-by: Anselm Fehnker <fehnker@fim.uni-passau.de>

Commit:
Adjust 'issue_processing.py' to the new GitHub-Wrapper format

The 'issue_processing.py' script is adjustet to the new GitHub-Wrapper
format. This means espacially changes in the information extraction of
related commits and related issues aswell as a new case for the
'unlabeled' event.

Signed-off-by: Anselm Fehnker <fehnker@fim.uni-passau.de>

ecklbarb commented on issue_processing/jira_issue_processing.py:
```diff
@@ -138,11 +173,12 @@ def merge_user_with_user_from_csv(user, persons):
 
-def parse_xml(issue_data, persons):
+def parse_xml(issue_data, persons, skip_history):
     """Parse issues from the xml-data
```
```
"""
Parse issues from the xml-data
```
Maybe, you can add a line break here to get consistent comments. Thank you, @fehnkera !

ecklbarb commented on issue_processing/jira_issue_processing.py:
```diff
@@ -168,65 +203,80 @@ def parse_xml(issue_data, persons):
-            user["name"] = reporter.firstChild.data
-            user["email"] = ""
-            issue["author"] = merge_user_with_user_from_csv(user, persons)
+            user = create_user(reporter.firstChild.data, reporter.attributes['username'].value, '')
```
It is a good idea to add the function `create_user`. :)

ecklbarb reviewed (commented):
Overall, its look really great. Thank you, @fehnkera!

clhunsen commented on issue_processing/issue_processing.py:
```diff
@@ -218,41 +218,38 @@ def merge_issue_events(issue_data):
+            if rel_commit["referenced_at"] is None:
+                rel_commit["user"] = create_user("", "", "")
+                rel_commit["created_at"] = ""
+                rel_commit["event"] = "hasCommit"
```
Is there a reason why this event is written in camelCase instead of snake_case?

fehnkera replied:
The camelCase name is the name @bockthom suggested. But I can change it to snake_case if preferred.

clhunsen replied:
I was just wondering because all other event names I found have a snake_case pattern (at least, in the overview file linked in the PR's first post). I am fine with the name, but I suggest to convert it to the snake_case variant `has_commit` for consistency. :wink:

clhunsen reviewed (approved):
Nice work, @fehnkera. This PR looks good to me, I just have one minor comment.

Commit:
Add None case to 'create_user' mehtod

The 'create_user' method in 'issue_processing' and
'jira_issue_processing' now also checks the none case as it could leed
to errors before.

Signed-off-by: Anselm Fehnker <fehnker@fim.uni-passau.de>

Commit:
Reformat code to a consistant format

The code in 'issue_processing.py' and 'jira_issue_processing.py' is
reformatted:
- both use double quotation marks
- line breaks in method headers are consistent
- event names are written in snake_case

Signed-off-by: Anselm Fehnker <fehnker@fim.uni-passau.de>

clhunsen commented:
Thanks for the tedious work on the last two commits, @fehnkera. 

For me, this PR looks fine now. @ecklbarb, do you have anything else to discuss or improve? I will wait for your approval before merging.

ecklbarb commented:
Thank you for the whole work you have done, @fehnkera.

For me, this PR looks fine, too. @clhunsen, you can merge the changes into the master.

bockthom commented:
@fehnkera Also a thank-you from mine. Good work! 
(Sorry for being inconsistent with my suggestion regarding "has_commits" using camel case, unfortunately I had not checked which coding style was used in the file before.)

Nevertheless, I have one question regarding the updated diagram:

> The updated diagram with the new events:
> [codeface-data-extraction-diagramm.pdf](https://github.com/se-passau/codeface-extraction/files/2353126/codeface-data-extraction-diagramm.pdf)

Does the diagram already comprise the new differentiation between referenced commits and the new "has_commits" event? At least, I cannot find this event in the updated diagram. Maybe it should not appear there and I am wrong or I have overlooked it, but if this event is missing there, please add it in the diagram! Thanks a lot!

fehnkera commented:
Thank you for your note @bockthom. You are right, the differentiation between referenced commits and "has_commits" was missing in the diagram. So here is an updated version with this event in it:
[codeface-data-extraction-diagramm.pdf](https://github.com/se-passau/codeface-extraction/files/2375135/codeface-data-extraction-diagramm.pdf)

bockthom commented:
Thanks for updating the diagram again. Looks good now!
